packages/network-controller/src/NetworkController.test.ts

```diff
@@ -1,224 +1,3653 @@
+                  const fakeInfuraProvider = buildFakeInfuraProvider();
+                  createInfuraProviderMock.mockReturnValue(fakeInfuraProvider);
+                  const fakeInfuraSubprovider = buildFakeInfuraSubprovider();
+                  SubproviderMock.mockReturnValue(fakeInfuraSubprovider);
+                  const fakeMetamaskProvider = buildFakeMetamaskProvider();
+                  createMetamaskProviderMock.mockReturnValue(
+                    fakeMetamaskProvider,
+                  );
+
+                  controller.providerConfig = {
+                    // NOTE: Neither the type nor chainId needs to match the
+                    // values in state, or match each other
+                    type: 'mainnet',
+                    chainId: '99999',
+                    nickname: 'some nickname',
+                  };
 
-  (
-    ['kovan', 'rinkeby', 'ropsten', 'mainnet', 'localhost'] as NetworkType[]
-  ).forEach((n) => {
-    it(`should create a provider instance for ${n} infura network`, () => {
-      const networkController = setupController(n, messenger);
-      expect(networkController.provider instanceof Web3ProviderEngine).toBe(
-        true,
+                  expect(createInfuraProviderMock).toHaveBeenCalledWith({
+                    network: networkType,
+                    projectId: 'infura-project-id',
+                  });
+                  expect(createMetamaskProviderMock).toHaveBeenCalledWith({
+                    type: 'mainnet',
+                    chainId: '99999',
+                    nickname: 'some nickname',
+                    dataSubprovider: fakeInfuraSubprovider,
+                    engineParams: {
+                      blockTrackerProvider: fakeInfuraProvider,
+                      pollingInterval: 12000,
+                    },
+                  });
+                  expect(controller.provider).toBe(fakeMetamaskProvider);
+                },
+              );
+            });
+
+            it('calls getEIP1559Compatibility', async () => {
+              await withController(
+                {
+                  state: {
+                    providerConfig: buildProviderConfig({
+                      type: networkType,
+                    }),
+                  },
+                  infuraProjectId: 'infura-project-id',
+                },
+                ({ controller }) => {
+                  const fakeInfuraProvider = buildFakeInfuraProvider();
+                  createInfuraProviderMock.mockReturnValue(fakeInfuraProvider);
+                  const fakeInfuraSubprovider = buildFakeInfuraSubprovider();
+                  SubproviderMock.mockReturnValue(fakeInfuraSubprovider);
+                  const fakeMetamaskProvider = buildFakeMetamaskProvider();
+                  createMetamaskProviderMock.mockReturnValue(
+                    fakeMetamaskProvider,
+                  );
+                  jest
+                    .spyOn(controller, 'getEIP1559Compatibility')
```

Comment:
Better to avoid mocking out controller methods, as this is part of the code under test.

Could we instead pass a stub into the mock provider engine, and check that the stub was called? I see that the fake provider engine class has that functionality already.

Reply:
I guess the same goes for all uses of `spyOn` in these tests. I see that there are a lot.

Reply:
I was trying to avoid adding more setup to ensure that `getEIP1559Compatibility` makes a request, but perhaps it's not that bad. I'll take a second look.

Reply:
I guess my comment didn't really explain much. It's true that it's the code under test, but my thought was, considering that we already test `getEIP1559Compatibility` on its own, do we care what it does exactly? That method is public, so we have the luxury of being able to stub it out. (The same thing goes for `lookupNetwork`.) I suppose this is the mockist vs. integrationist argument. I tend to be okay with the former but I can see a case for the latter. I'll try it out.

Reply:
Most of the same problems with stubbing internal methods still apply when stubbing public ones.

* It makes the tests dependent upon that specific function signature, so the tests might break even if the thing-under-test hasn't been broken.
* If this function was used incorrectly (i.e. wrong parameters passed in, called at the wrong time, return value ignored, , something like that), this test wouldn't catch it.

I guess this is down to mockist v.s. classicist again. I'm still having trouble wrapping my head around why this style of test would be helpful for a project. It seems equivalent to a careful audit, in that testing a function like this entails carefully examining it. But less mocking would make for a better, more accurate audit. And the test we merge in that gets run on each PR provides more friction than benefit on an onging basis if it's written like this.

Reply:
Not exactly  - what I had in mind was testing that it has the intended effect. e.g. change this description to something like "checks EIP 1559 compatibility", and check the controller state to ensure the EIP 1559 compatibility for this network was set correctly.

By testing the intended effect, the test is more robust in that refactors won't break it, and we also cover that the "glue" is setup correctly (if it wasn't, it wouldn't have the intended effect).

Reply:
Oh, of course. That would make a lot more sense 🤦🏻

Reply:
Okay, I've gone through most tests that stub `lookupNetwork` and `getEIP1559Compatibility` so that they call these methods for real. There are some tests that I left alone, however:

* There are tests which involve an instance of EthQuery somehow not having a `sendAsync` method. I think the best way to do this is to make a fake EthQuery object. 
* There are also tests for `lookupNetwork` or `getEIP1559Compatibility` themselves. Now in order to use either of these methods, you have to set a provider first. Of course, in the process of setting a provider, both of these methods end up getting called, so if we are testing a specific piece of behavior within one of these methods, it would end up getting called twice, and it would introduce more overhead because it may create more state changes that aren't really relevant to the test itself. But, maybe we want to allow these too?

Anyway, you can see new changes here: [1b2cebb](https://github.com/MetaMask/core/pull/1039/commits/1b2cebbb6299914e1ab92e13d93108949ddafbb7). Happy to make more changes if you think we need them.